From a5450e840a901281de50408d4ce3233a4fb166e0 Mon Sep 17 00:00:00 2001
From: Alison Wang <b18965@freescale.com>
Date: Wed, 13 May 2015 09:46:13 +0800
Subject: [PATCH 844/845] arm: ls1021a: add platform notifier for dma-coherent
 requirement

This patch adds platform notifier for dma-coherent requirement.
Structure arm_coherent_dma_ops is used instead of arm_dma_ops.

Signed-off-by: Alison Wang <alison.wang@freescale.com>
Change-Id: I55e755b5ead1cb50ff9ca2e6838588a04e5ea9e7
Reviewed-on: http://git.am.freescale.net:8181/36291
Tested-by: Review Code-CDREVIEW <CDREVIEW@freescale.com>
Reviewed-by: Claudiu Manoil <claudiu.manoil@freescale.com>
Reviewed-by: Zhengxiong Jin <Jason.Jin@freescale.com>
[Xulin:Original patch taken from
Freescale-Linux-SDK-for-LS1021A-IOT-Rev2-v0.4-SOURCE-20150907-yocto.iso, and
distinguish the Rev1 & Rev2 IOT board.]
Signed-off-by: Xulin Sun <xulin.sun@windriver.com>
---
 arch/arm/mach-imx/mach-ls1021a.c | 30 ++++++++++++++++++++++++++++++
 1 file changed, 30 insertions(+)

diff --git a/arch/arm/mach-imx/mach-ls1021a.c b/arch/arm/mach-imx/mach-ls1021a.c
index 3222694..a0c15e9 100644
--- a/arch/arm/mach-imx/mach-ls1021a.c
+++ b/arch/arm/mach-imx/mach-ls1021a.c
@@ -8,11 +8,40 @@
  */
 
 #include <asm/mach/arch.h>
+#include <linux/dma-mapping.h>
 #include <linux/of_platform.h>
+#include <linux/platform_data/dcfg-ls1021a.h>
 #include <linux/phy.h>
 #include <linux/phy_fixed.h>
 #include "common.h"
 
+static int ls1021a_platform_notifier(struct notifier_block *nb,
+				  unsigned long event, void *__dev)
+{
+	struct device *dev = __dev;
+
+	if (event != BUS_NOTIFY_ADD_DEVICE)
+		return NOTIFY_DONE;
+
+	if (of_device_is_compatible(dev->of_node, "fsl,etsec2"))
+		set_dma_ops(dev, &arm_coherent_dma_ops);
+	else if (of_property_read_bool(dev->of_node, "dma-coherent"))
+		set_dma_ops(dev, &arm_coherent_dma_ops);
+
+	return NOTIFY_OK;
+}
+
+static struct notifier_block ls1021a_platform_nb = {
+	.notifier_call = ls1021a_platform_notifier,
+};
+
+static void __init ls1021a_init_machine(void)
+{
+	if (!is_ls1021a_rev1())
+		bus_register_notifier(&platform_bus_type, &ls1021a_platform_nb);
+	of_platform_populate(NULL, of_default_bus_match_table, NULL, NULL);
+}
+
 #ifdef CONFIG_FIXED_PHY
 static int __init of_add_fixed_phys(void)
 {
@@ -51,5 +80,6 @@ static const char * const ls1021a_dt_compat[] __initconst = {
 
 DT_MACHINE_START(LS1021A, "Freescale LS1021A")
 	.smp		= smp_ops(ls1021a_smp_ops),
+	.init_machine   = ls1021a_init_machine,
 	.dt_compat	= ls1021a_dt_compat,
 MACHINE_END
-- 
2.0.2

